include/nbl/system/demote_promote_writer_readers_lock.h

```diff
@@ -193,86 +164,122 @@ class demote_promote_writer_readers_lock
 				if (!registeredPending)
 				{
 					registeredPending = true;
-					nextState += impl::DPWRLStateSemantics{ .currentReaders = 0, .pendingWriters = 0, .pendingUpgrades = 1, .writing = false, .stateLocked = false };
+					preemptedState += impl::DPWRLStateSemantics{ .currentReaders = 0, .pendingWriters = 0, .pendingUpgrades = 1, .writing = false, .stateLocked = false };
+					preemptedState -= impl::DPWRLStateSemantics{ .currentReaders = 1, .pendingWriters = 0, .pendingUpgrades = 0, .writing = false, .stateLocked = false };
 				}
-				// Release flipLock
-				state.store(nextState);
-				if (spinCount > SpinsBeforeYield)
-					std::this_thread::yield();
-			}
-			// Release flipLock, declare that this thread is writing, mark that this thread is no longer pending IF had been registered as such
-			if (registeredPending)
-				nextState -= impl::DPWRLStateSemantics{ .currentReaders = 0, .pendingWriters = 0, .pendingUpgrades = 1, .writing = false, .stateLocked = false };
-			nextState |= writingMask;
-			state.store(nextState);
-			break;
-		}
+				return preemptedState;
+			};
+
+		perform_under_locked_state(std::move(success), preemptionCheck, sanityChecks, preempted);
 	}
 
 	/**
 	* @brief Downgrades a thread with a writer's lock to a reader's lock. If a thread already has a writer's lock, use this instead of `write_unlock()` -> `read_lock()`
 	*/
 	void downgrade()
 	{
-		for (state_lock_value_t spinCount = 0; true; spinCount++)
+		const auto success = [](const state_lock_value_t oldState) -> state_lock_value_t
+			{
+				// Declare that thread is no longer writing and that it will be reading
+				state_lock_value_t newState = oldState & ~writingMask;
+				newState += impl::DPWRLStateSemantics{ .currentReaders = 1, .pendingWriters = 0, .pendingUpgrades = 0, .writing = false, .stateLocked = false };
+				return newState;
+			};
+
+		const auto sanityChecks = [&](const state_lock_value_t oldState) -> bool
+			{
+				// If we had a writer's lock, the writing flag should be on
+				// If we had a writer's lock, no thread should be currently reading
+				return (oldState & writingMask) && !(oldState & currentReadersMask);
+			};
+
+		perform_under_locked_state(std::move(success), defaultPreemptionCheck, sanityChecks);
+	}
+
+private:
+
+	constexpr static auto defaultPreemptionCheck = [](const state_lock_value_t oldState) -> bool {return false; };
+	constexpr static auto defaultPreempted = [](const state_lock_value_t oldState)->state_lock_value_t {return oldState; };
+	constexpr static auto defaultSanityChecks = [](const state_lock_value_t oldState)->bool {return true; };
+
+	template<typename Success, typename PreemptionCheck = decltype(defaultPreemptionCheck), typename SanityChecks = decltype(defaultSanityChecks), typename Preempted = decltype(defaultPreempted)>
+	inline void perform_under_locked_state(
+		Success&& success,
+		PreemptionCheck& preemptionCheck = defaultPreemptionCheck,
```

Comment:
don't make `preemptionCheck` stateful then, just take it as a `typename` and call `operator()` on a fresh object